### PR TITLE
Issue 4509: Rack-aware Bookkeeper client config

### LINKED ISSUE #4509: Explore rack-aware data placement to write to Bookkeeper
**Problem description**
Currently, the Segment Store writes to Bookkeeper using the default placement policy (i.e., get random Bookies). While this may be acceptable if Bookies use protected volumes for data storage, it is clear that in the case of using local drives it may lead to data loss. That is, in the case of running 9 Bookies on 3 physical nodes, and setting in the Segment Store a writeQuorum=3, it could be the case that the 3 replicas for a write are written to 3 Bookies on the same physical node. If the node crashes, we cannot recover this data.

For this reason, we need to explore the data placement strategies that the Bookkeeper client offers to prevent or minimize data loss probabilities.

**Problem location**
Bookkeeper client config.

**Suggestions for an improvement**
Explore the use of `RackawareEnsemblePlacementPolicy`.

### PR DESCRIPTION
**Pull request by RaulGracia**
**Change log description**  
Allow the Bookkeeper client in the Segment Store to be configured with rack-aware ensemble placement policy.

**Purpose of the change**  
Fixes #4509.

**What the code does**  
This PR enables the Bookkeeper client in the Segment Store to be configured with a rack-aware ensemble placement policy. To this end, we introduce 3 new Bookkeeper client parameters:
- `bookkeeper.enforceMinNumRacksPerWriteQuorum`: This parameter defines whether we should enforce the rack-aware policy or not.
- `bookkeeper.minNumRacksPerWriteQuorum`: In the case we enforce rack-aware ensamble placement policy, this parameter determines the number of different physical location that a write should be replicated to.
- `bookkeeper.networkTopologyScriptFileName`: This is an important and environment-dependent parameter. It should point to a script that, based on a list of Bookie IPs as input, outputs a list of equal size with the physical locations of each Bookie IP (e.g. `[IP1, IP2, IP3]` -> `[/rack1, /rack2, /rack3]`). We provide a sample file that is intended to be used in test or demo situations only.

**How to verify it**  
We have deployed Pravega with the changes of this branch. As can be see, the `RackawareEnsemblePlacementPolicy` is properly initialized:
```
2020-01-22 18:29:00,845 960  [main] INFO  o.a.b.c.RackawareEnsemblePlacementPolicyImpl - Initialize rackaware ensemble placement policy @ <Bookie:172.28.13.11:0> @ /datacenter/rack-11 : org.apache.bookkeeper.net.ScriptBasedMapping.
```
This means that the Bookkeeper client has correctly initialized the `ScriptBasedMapping`, which internally invokes the topology script (`networkTopologyScriptFileName`).

Then, upon each Bookie join, the network topology is updated correctly, which implies that the script is correctly invoked:
``` 
2020-01-22 18:29:00,873 988  [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.c.RackawareEnsemblePlacementPolicyImpl - Cluster changed : left bookies are [], joined bookies are [pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181, pravega-bookie-2.pravega-bookie-headless.default.svc.cluster.local:3181, pravega-bookie-0.pravega-bookie-headless.default.svc.cluster.local:3181], while dead bookies are [].
2020-01-22 18:29:00,877 992  [BookKeeperClientScheduler-OrderedScheduler-0-0] INFO  o.a.b.net.NetworkTopologyImpl - Adding a new node: /datacenter/rack-10/pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181
2020-01-22 18:29:00,877 992  [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.net.NetworkTopologyImpl - NetworkTopology became:
/datacenter/rack-10/pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181
2020-01-22 18:29:00,878 993  [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.c.RackawareEnsemblePlacementPolicyImpl - Cluster changed : bookie pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181 joined the cluster.
2020-01-22 18:29:00,881 996  [BookKeeperClientScheduler-OrderedScheduler-0-0] INFO  o.a.b.net.NetworkTopologyImpl - Adding a new node: /datacenter/rack-9/pravega-bookie-2.pravega-bookie-headless.default.svc.cluster.local:3181
2020-01-22 18:29:00,881 996  [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.net.NetworkTopologyImpl - NetworkTopology became:
/datacenter/rack-10/pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181
/datacenter/rack-9/pravega-bookie-2.pravega-bookie-headless.default.svc.cluster.local:3181
2020-01-22 18:29:00,881 996  [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.c.RackawareEnsemblePlacementPolicyImpl - Cluster changed : bookie pravega-bookie-2.pravega-bookie-headless.default.svc.cluster.local:3181 joined the cluster.
2020-01-22 18:29:00,884 999  [BookKeeperClientScheduler-OrderedScheduler-0-0] INFO  o.a.b.net.NetworkTopologyImpl - Adding a new node: /datacenter/rack-8/pravega-bookie-0.pravega-bookie-headless.default.svc.cluster.local:3181
2020-01-22 18:29:00,884 999  [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.net.NetworkTopologyImpl - NetworkTopology became:
/datacenter/rack-10/pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181
/datacenter/rack-9/pravega-bookie-2.pravega-bookie-headless.default.svc.cluster.local:3181
/datacenter/rack-8/pravega-bookie-0.pravega-bookie-headless.default.svc.cluster.local:3181
2020-01-22 18:29:00,884 999  [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.c.RackawareEnsemblePlacementPolicyImpl - Cluster changed : bookie pravega-bookie-0.pravega-bookie-headless.default.svc.cluster.local:3181 joined the cluster.
2020-01-22 18:29:00,885 1000 [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.c.RackawareEnsemblePlacementPolicyImpl - Cluster changed : left bookies are [], joined bookies are [], while dead bookies are [].
```
As you can see, the final topology relates each Bookie to a different rack number, that corresponds to the last number of the Bookie IP (as explained in the sample Bookkeeper topology script):
```
pravega-bookie-0   1/1     Running   1          13m     xxx.8
pravega-bookie-1   1/1     Running   0          13m     yyy.10
pravega-bookie-2   1/1     Running   0          13m     zzz.9
```
```
2020-01-22 18:29:00,884 999  [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.net.NetworkTopologyImpl - NetworkTopology became:
/datacenter/rack-10/pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181
/datacenter/rack-9/pravega-bookie-2.pravega-bookie-headless.default.svc.cluster.local:3181
/datacenter/rack-8/pravega-bookie-0.pravega-bookie-headless.default.svc.cluster.local:3181
```

### CODE REVIEW

File: docker/pravega/scripts/sample-bookkeeper-topology.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env sh
+#
+# Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
```
Comment (andreipaduroiu) [inline docker/pravega/scripts/sample-bookkeeper-topology.sh]: Please remove year.

File: gradle.properties [hunk 2]
```diff
@@ -16,7 +16,7 @@ apacheCommonsCompressVersion=1.18
 apacheCuratorVersion=4.0.1
 apacheZookeeperVersion=3.5.5
 checkstyleToolVersion=8.23
-bookKeeperVersion=4.7.3
+bookKeeperVersion=4.9.2
```
Comment (andreipaduroiu) [inline gradle.properties]: I think this may be unintended

Comment (RaulGracia) [reply gradle.properties]: With the client `4.7.3` I experienced many issues, and as soon as I upgraded it worked fine. As this is going to `master`, we can wait until merging #4394.

File: docker/pravega/scripts/sample-bookkeeper-topology.sh (hunk 1 above)
Comment (RaulGracia) [reply docker/pravega/scripts/sample-bookkeeper-topology.sh]: Ok.

**Comment by RaulGracia** [comment]
@andreipaduroiu please, go ahead and merge this PR. We are having problems with the `4.9.2` version of the Bookkeeper client, so this PR upgrades it to the latest major version that does not bring additional problems (e.g., build failures, performance issues). Upgrading to 4.9.2 will require additional work.

**Pull request merged**